### PR TITLE
Loudness normalization with `pyloudnorm`

### PR DESCRIPTION
**Pull request by desh2608**
Related to https://github.com/lhotse-speech/lhotse/issues/966.

This PR adds a method `normalize_loudness()` for recordings. This takes an argument `target` which specifies the desired loudness (usually around -23 dB is a good value). The implementation uses [pyloudnorm](https://github.com/csteinmetz1/pyloudnorm/tree/master).

Also, we move `ReverbWithImpulseResponse` out of the torchaudio.py and into its own file, since it is not a torchaudio based transform.

### CODE REVIEW

File: lhotse/augmentation/loudness.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    meter = pyln.Meter(
+        sampling_rate, block_size=min(0.4, duration)
+    )  # create BS.1770 meter
+    loudness = meter.integrated_loudness(audio.T)
+
+    # loudness normalize audio to target LUFS. We will ignore the warnings related to
+    # clipping the audio.
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        loudness_normalized_audio = pyln.normalize.loudness(audio.T, loudness, target)
```
Comment (pzelasko) [inline lhotse/augmentation/loudness.py:70]: If audio clipping is an issue here you can add a limiter as a post-processing step, e.g. https://github.com/pzelasko/cylimiter

**Review by pzelasko — APPROVED** [review]
Looks good to me, I left a single comment, up to you :)

File: lhotse/augmentation/loudness.py (hunk 1 above)
Comment (desh2608) [reply lhotse/augmentation/loudness.py:70]: Sure, I can add it. But I don't have enough familiarity with limiters (or "loudness" for that matter) to know what to do exactly.

Comment (pzelasko) [reply lhotse/augmentation/loudness.py:70]: The defaults should "just work" with pretty much anything, it basically keeps track of the signal's loudness with a small lookahead and reduces the gain if it crosses some threshold. Think of it as a soft clipping that doesn't introduce as much distortion as hard clipping.

Comment (pzelasko) [reply lhotse/augmentation/loudness.py:70]: .. but again I don't know if that's a real problem with this approach and worth the extra dependency, so it's your call :)

Comment (desh2608) [reply lhotse/augmentation/loudness.py:70]: So far I have only used it to make the LibriCSS distant mic audio louder (it is at around -52 dB originally), and it sounds okay even with the clipping. I suppose we can let it be for now and add the limiter later if someone needs it?

File: lhotse/augmentation/loudness.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+
+@dataclass
+class LoudnessNormalization(AudioTransform):
+    """
+    Loudness normalization based on pyloudnorm: https://github.com/csteinmetz1/pyloudnorm.
+    """
+
+    target: float
+    sampling_rate: int = 16000
```
Comment (lifeiteng) [inline lhotse/augmentation/loudness.py:19]: @pzelasko should we make `sampling_rate` as a member?

```
class AudioTransform:
    def __call__(self, samples: np.ndarray, sampling_rate: int) -> np.ndarray:
        """
        Apply transform.

        To be implemented in derived classes.
        """
        raise NotImplementedError
```

File: lhotse/audio.py [hunk 3]
```diff
@@ -707,6 +708,23 @@ def perturb_volume(self, factor: float, affix_id: bool = True) -> "Recording":
...
+        """
+        Return a new ``Recording`` that will lazily apply WPE dereverberation.
+
+        :param target: The target loudness (in dB) to normalize to.
+        :param affix_id: When true, we will modify the ``Recording.id`` field
+            by affixing it with "_ln{factor}".
+        :return: a modified copy of the current ``Recording``.
+        """
+        transforms = self.transforms.copy() if self.transforms is not None else []
+        transforms.append(LoudnessNormalization(target=target).to_dict())
```
Comment (lifeiteng) [inline lhotse/audio.py:721]: `sampling_rate` is missed

**Comment by lifeiteng** [comment]
@desh2608 @pzelasko 

I used set.normalize_loudness as blow: https://github.com/lifeiteng/vall-e/blob/main/valle/bin/tokenizer.py#L173
```
if args.prefix == "aishell":
    # NOTE: the loudness of aishell audio files is around -33
    # The best way is datamodule --on-the-fly-feats --enable-audio-aug
    cut_set = cut_set.normalize_loudness(
        target=-20.0, affix_id=True
    )

cut_set = cut_set.resample(24000)
```

But model's accuracy drops a lot.


截屏2023-04-14 23 44 31

**Comment by desh2608** [comment]
@lifeiteng what metric is this curve? If it is validation accuracy, did you normalize loudness for both train and validation sets? I would think that for the training set, it may be better to "perturb" the loudness (within some range) rather than normalize it, similar to how we perturb volume.

**Comment by lifeiteng** [comment]
@desh2608 Training CrossEntropy ArTop10Accuracy.
yes, normalized loudness for both train and validation sets.
It's used in text-to-speech, yes "perturb" is better, but I want to make it simple at current stage.
I don't understand why normalization leads to a significant drop in accuracy. The bug(comments above) is not triggered.

**Comment by pzelasko** [comment]
@lifeiteng If it's for TTS, can you listen to the train and dev examples before and after normalization, as well to the model predictions? Maybe that could reveal if there's something funky going on.
